File: iceoryx2-ffi/ffi/src/api/message_type_details.rs
// Copyright (c) 2024 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

use core::ffi::c_char;

use iceoryx2::service::static_config::message_type_details::*;

use crate::{iox2_type_variant_e, IOX2_TYPE_NAME_LENGTH};

#[derive(Clone, Copy)]
#[repr(C)]
pub struct iox2_type_detail_t {
    pub variant: iox2_type_variant_e,
    pub type_name: [c_char; IOX2_TYPE_NAME_LENGTH],
    pub size: usize,
    pub alignment: usize,
}

#[derive(Clone, Copy)]
#[repr(C)]
pub struct iox2_message_type_details_t {
    pub header: iox2_type_detail_t,
    pub user_header: iox2_type_detail_t,
    pub payload: iox2_type_detail_t,
}

impl From<&TypeDetail> for iox2_type_detail_t {
    fn from(value: &TypeDetail) -> Self {
        Self {
            variant: (&value.variant()).into(),
            type_name: core::array::from_fn(|n| {
                if n < value.type_name().len() {
                    value.type_name().as_bytes()[n] as _
                } else {
                    0
                }
            }),
            size: value.size(),
            alignment: value.alignment(),
        }
    }
}

impl From<&MessageTypeDetails> for iox2_message_type_details_t {
    fn from(m: &MessageTypeDetails) -> Self {
        Self {
            header: (&m.header).into(),
            user_header: (&m.user_header).into(),
            payload: (&m.payload).into(),
        }
    }
}

File: iceoryx2-cxx/src/message_type_details.cpp
// Copyright (c) 2024 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

#include "iox2/message_type_details.hpp"
#include "iox/into.hpp"

namespace iox2 {
TypeDetail::TypeDetail(iox2_type_detail_t value)
    : m_value { value } {
}

auto TypeDetail::variant() const -> TypeVariant {
    return iox::into<TypeVariant>(static_cast<int>(m_value.variant));
}

auto TypeDetail::type_name() const -> const char* {
    return &m_value.type_name[0];
}

auto TypeDetail::size() const -> size_t {
    return m_value.size;
}

auto TypeDetail::alignment() const -> size_t {
    return m_value.alignment;
}

MessageTypeDetails::MessageTypeDetails(iox2_message_type_details_t value)
    : m_value { value } {
}

auto MessageTypeDetails::header() const -> TypeDetail {
    return TypeDetail(m_value.header);
}

auto MessageTypeDetails::user_header() const -> TypeDetail {
    return TypeDetail(m_value.user_header);
}

auto MessageTypeDetails::payload() const -> TypeDetail {
    return TypeDetail(m_value.payload);
}
} // namespace iox2
